# script to categorize reads based on alignment results
# input = report text files and sample name (e.g. REPORTS/SH-1_ABAB_2_report.txt and SH-1) 
# output = report text files with categories (will add new columns to current report file e.g. REPORTS/SH-1_ABAB_2_report.txt)
 
import os, sys
import pandas as pd
import pysam
import re

fn = sys.argv[1]
sample = sys.argv[2]

def read_fasta(reporter):
    seq = ''
    with open(f'PTREseq_ref_seq/{reporter}.fa', 'r') as fn:
        for line in fn:
             if not line.startswith('>'):
                 seq += line.rstrip()

    return seq

def get_correct_query(read, bam):
    for query in bam.fetch(until_eof=True):
        if read in query.query_name:
            cigar = query.cigarstring
            if cigar is None:
                cigar = 'error_cigar'
        
            try:
                tag = query.get_tag('NM')
            except:
                tag = 'error_query'
            
            return cigar, tag

    return 'error_read_not_found', 'error_read_not_found'

def categorize_read(cigar, dist, ref_seq):
    cigar_tuple = re.findall(r'(\d+)([A-Z]|={1})', cigar)
    total_insert = 0
    AG_deletion = 0
    pos_in_ref = 0

    for length, operation in cigar_tuple:
        if operation == '=' or operation == 'M':
            pos_in_ref += int(length)
        elif operation == 'D':
            if int(length) > 1:
                start = pos_in_ref
                start_seq = ref_seq[start:start+2]
                end = pos_in_ref + int(length)
                end_seq = ref_seq[end-2:end]
                if start_seq == 'GT' and end_seq == 'AG':
                    AG_deletion += int(length)
                    donor = ref_seq[start-6:start]
                    donor_loc = start 
                    acceptor = ref_seq[end:end+6]
                    acceptor_loc = end
            pos_in_ref += int(length)
        elif operation == 'I' and int(length) > 1:
            total_insert += int(length)
        elif operation == 'S':
            pass

    if total_insert > 5:
        return 'too_many_inserts'
    elif AG_deletion > 0:
        new_dist = dist - AG_deletion
        junction = f'{donor}_{donor_loc}_{acceptor}_{acceptor_loc}'
        if new_dist < 20:
            return f'AG_junction_low_mismatch-{junction}'
        else:
            return f'AG_junction_high_mismatch-{junction}'
    else:
        if dist < 20:
            return 'full_length'
        else:
            return 'ambiguous'

def get_distances(read_name, read_type, reporters, _cigar, _dist, ref_seq):
    if read_type == 'perfect_match' or read_type == 'one_mismatch':
        if _cigar is None or _dist is None:
            category = 'one_more_errors' 
        else:    
            category = categorize_read(_cigar, _dist, ref_seq)

        cigar = _cigar
        dist = _dist
    else:
        cigar = read_type
        dist = read_type
        category = read_type
    return cigar, dist, category

linestowrite = []
with open(f'REPORTS/{fn}', 'r') as report:
	if 'error' in fn:
		# for reads with 1 of the 3 errors, simply add these errors as new columns
		for line in report:
			spl = line.rstrip().split('\t')
			error = spl[2]
			towrite = f'{line.rstrip()}\t{error}\t{error}\t{error}\n'
			linestowrite.append(towrite)	
	else:
		reporter = fn.split('_')[1] + '_' + fn.split('_')[2]
		# alignment file
		bam_file = f'ALIGNMENT/{sample}_{reporter}_demux.sam.bam'
		#index_file = f'ALIGNMENT/{sample}_{reporter}_demux.sam.bam.sorted.bam.bai'  
		bam = pysam.AlignmentFile(bam_file, 'r')
		bam_dict = {}
		for query in bam.fetch(until_eof = True):
			name = query.query_name.split(' ')[0]
			cigar = query.cigarstring
			try:
				tag = query.get_tag('NM')
			except:
				tag = None
			bam_dict[name] = (cigar, tag)	
			

		# reference sequence
		ref_seq = read_fasta(reporter)    
		
		for line in report:	
			spl = line.rstrip().split('\t')
			original_line = '\t'.join(spl[:4])
			_cigar, _dist = bam_dict[spl[0]]
			if _cigar is None or _dist is None:
				cigars, distances, category = ['error', 'error', 'error']
			else:	
				cigars, distances, category = get_distances(spl[0], spl[2], spl[3], _cigar, _dist, ref_seq)
			
			# add 4 new columns: cigar string, edit distance, category, and reporter
			towrite = f'{original_line.rstrip()}\t{cigars}\t{distances}\t{category}\t{reporter}\n'	
			linestowrite.append(towrite)	

with open(f'REPORTS/{fn}', 'w') as out:
	for line in linestowrite:
		out.write(line)
